File: 8week/seungboshim/부대복귀.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<vector<int>> makeGraph(int n, vector<vector<int>> roads) {
    vector<vector<int>> graph(n);
    for(auto& road: roads) {
        int u = road[0];
        int v = road[1];
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    
    return graph;
}

int BFS(int start, int dest, vector<vector<int>> graph) {
    vector<int> dist(graph.size(), -1);
    queue<int> q;
    
    q.push(start);
    dist[start] = 0;
    
    while (!q.empty()) {
        int current = q.front();
        q.pop();
        
        for (int n: graph[current]) {
            if (dist[n] == -1) {
                q.push(n);
                dist[n] = dist[current] + 1;
                
                if (n == dest) {
                    return dist[n];
                }
            }
        }
    }
    
    return -1;
}

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
    vector<int> answer;
    vector<vector<int>> graph = makeGraph(n, roads);
    
    for(int s: sources) {
        // s -> dest 그래프 최단거리
        int result = BFS(s, destination, graph);
        answer.push_back(result);
    }
    
    return answer;
}

File: 8week/cokemania2/부대복귀.py
from collections import deque

def create_map_dict(n):
    """
    모든 맵 정보를 -1로 초기화하여 딕셔너리를 생성
    """
    return {i: -1 for i in range(1, n + 1)}

def save_short_cuts(destination, map_dict, road_dict):
    """
    BFS 방식으로 최단 거리를 계산하여 딕셔너리에 저장
    도달하는 지역이 -1(미방문)이거나 더 빨리 방문 가능하면 진행
    """
    queue = deque([[destination, 0]])
    
    while queue:
        road, count = queue.popleft()
        if map_dict[road] == -1 or map_dict[road] > count:
            map_dict[road] = count
        
            next_roads = road_dict[road]
            for next_road in next_roads:
                queue.append([next_road, count + 1])
            

def road_to_dict(a, b, dic):
    """
    경로 정보를 dict로 변환
    """
    if a in dic:
        dic[a].append(b)
    else:
        dic[a] = [b]

def solution(n, roads, sources, destination):
    answer = []
    road_dict = dict()
    
    for road in roads:
        road_to_dict(road[0], road[1], road_dict)
        road_to_dict(road[1], road[0], road_dict)
    
    map_dict = create_map_dict(n)
    save_short_cuts(destination, map_dict, road_dict)
    
    for source in sources:
        answer.append(map_dict[source])
    
    return answer